### DISCUSSION TITLE
To skip a specific service in trivy aws scan

### DISCUSSION THREAD

**Discussion by pritam-yadu** [aquasecurity/trivy#4287] [Q&A]
Trivy aws scan seems to crash intermittently while scanning workspaces service. Is there a way to skip specific service while scanning?

```
trivy aws --region us-east-2  --service workspaces --debug

Error occurred while running adapter for workspaces: operation error WorkSpaces: DescribeWorkspaces, exceeded maximum number of attempts, 3, https response error StatusCode: 0, RequestID: , request send failed, Post "https://workspaces.us-east-2.amazonaws.com/": dial tcp: lookup workspaces.ap-east-1.amazonaws.com on 172.18.0.10:53: no such h2023-01-12T16:45:32.668+0530
```

**Comment by simar7** [comment]
hi @pritam-yadu thanks for the report - can you share more detailed logs (preferably using the --debug flag) so we can see where it is crashing?

**Comment by simar7** [comment]
Going to close this as no further info has been provided. Please reopen/comment if you still see the issue.

**Comment by itaysk** [comment]
I've reopened since the question wasn't about the crash but about skipping the service
(separately, we should make sure we don't crash in case of a specific service error)

**Reply by pritam-yadu** [reply]
You are right. I wasn't reporting an issue. I was enquiring if there is any feature to skip specific service while scanning. This could be useful as some organizations blocks some services in SCP. Also can be useful as not all AWS services are supported in all regions.

**Comment by AnaisUrlichs** [comment]
Hi there, we currently do not have the option on the AWS scan to skip services (the only workaround would be putting services that you don't want to scan into a different region which doesn't make much sense 😆 ) so I assume this is a feature request

currently, we only have a way to specify specific service
```
--service strings   Only scan AWS Service(s) specified with this flag. Can specify multiple services using --service A --service B etc.
```

We could introduce a `--skip-service` or similar

**Reply by pritam-yadu** [reply]
Hi @AnaisUrlichs, yup. This is more of a feature request. --skip-service flag would be helpful for many use cases as I mentioned in previous comment.

**Comment by simar7** [comment]
Opened https://github.com/aquasecurity/trivy/issues/4619 to track the feature request.


@pritam-yadu: As for the error, could you provide more details as to what exactly happens? You mentioned that Trivy errors out but does it go ahead with other services being scanned? If not, we need to make that experience better.

Having any debug logs will greatly help.

**Reply by pritam-yadu** [reply]
@simar7 It is strange that I am not able to reproduce the error in my local; I was able to few days ago. But I am getting errors in production. I will try to explain what's going on in their with another example.

When I run trivy scan in eu-central-2, scan fails as ApiGatewayV2 from api-gateway service is not yet officially supported by aws in that region and I get below error. 

**command** : `trivy aws --region eu-central-2 --update-cache -f json`

**Result** : `b\'{\\n  "ArtifactName": "760********3",\\n  "ArtifactType": "aws_account",\\n  "Metadata": {\\n    "ImageConfig": {\\n      "architecture": "",\\n      "created": "0001-01-01T00:00:00Z",\\n      "os": "",\\n      "rootfs": {\\n        "type": "",\\n        "diff_ids": null\\n      },\\n      "config": {}\\n    }\\n  }\\n}\\n\'\n b"2023-06-16T05:17:02.473Z\\t\\x1b[33mWARN\\x1b[0m\\tAdapter error: failed to run adapter for api-gateway: operation error ApiGatewayV2: GetDomainNames, https response error StatusCode: 400, RequestID: 5b101a54-9e8f-424c-87a4-f946ab96ac1a, deserialization failed, failed to decode response body, invalid character \'\\\\\'\' in string escape code\\n"`

For problematic services like this, it would be nice to have --skip-service flag supported in trivy. 
It would also be nice if Trivy handles this service specific errors internally and continue scanning rest of the services instead of exiting the entire scan. This service specific errors then can be logged to  output. 
I know there is **--service** flag which I can use where I can provide all services that I want to scan but then I have to provide all those services in the command, also keep the track of services supported by Trivy and not supported by my organization and keep updating the code accordingly every time.

I am not sure how much this helps to you but I will give you debug logs I am able to reproduce this locally.
